### PR TITLE
Introduce `ReasonedRuleViolation` type to associate a reason with a violation position

### PR DESCRIPTION
**Pull request by SimplyDanny**
With this new type we can get rid of most of the custom `StyleViolation` creators.

### CODE REVIEW

File: Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift [hunk 1]
```diff
@@ -48,16 +49,49 @@ public extension SwiftSyntaxRule {
...
+    }
+
+    public static func < (lhs: Self, rhs: Self) -> Bool {
+        lhs.position < rhs.position
+    }
+}
+
+/// Extension for arrays of `ReasonedRuleViolation`s that provides the automatic conversion of
+/// `AbsolutePosition`s into `ReasonedRuleViolation`s (without a specific reason).
+extension Array where Element == ReasonedRuleViolation {
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift:81]: nice! i was looking into doing something similar, but this was the missing piece. I went with the protocol/associated type route and then faced a bunch of issues, so this is really clever ✨

File: Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift [hunk 2]
```diff
@@ -48,16 +49,49 @@ public extension SwiftSyntaxRule {
...
+    mutating func append(_ position: AbsolutePosition) {
+        append(ReasonedRuleViolation(position: position))
+    }
+
+    mutating func append(contentsOf positions: [AbsolutePosition]) {
+        positions.forEach { append($0) }
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift:87]: My gut feeling is that `append(contentsOf: positions.map {  ReasonedRuleViolation(position: $0) })` would be faster

File: Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift [hunk 3]
```diff
@@ -48,16 +49,49 @@ public extension SwiftSyntaxRule {
...
+            .map { makeViolation(file: file, violation: $0) }
+    }
+}
+
+/// A violation produced by `ViolationsSyntaxVisitor`s.
+public struct ReasonedRuleViolation: Comparable {
+    /// The violation's position.
+    public let position: AbsolutePosition
+    /// A specific reason for the violation.
+    public let reason: String?
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift:63]: while we're at this, can we add the `severity`? this is the other reason rules can't use `ViolationsSyntaxVisitor` currently

File: Source/SwiftLintFramework/Rules/Idiomatic/NoFallthroughOnlyRule.swift [hunk 4]
```diff
@@ -24,7 +24,7 @@ private extension NoFallthroughOnlyRule {
         override func visitPost(_ node: SwitchCaseListSyntax) {
             let cases = node.compactMap { $0.as(SwitchCaseSyntax.self) }
 
-            let violations = cases.enumerated()
+            cases.enumerated()
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Idiomatic/NoFallthroughOnlyRule.swift:27]: nit: can we revert this change? i feel like it's slightly easier to debug/read

File: Source/SwiftLintFramework/Rules/Style/ClosureParameterPositionRule.swift [hunk 5]
```diff
@@ -117,16 +117,15 @@ private extension ClosureParameterPositionRule {
                 return
             }
 
-            let violations = signature.positionsToCheck
+            signature.positionsToCheck
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Style/ClosureParameterPositionRule.swift:120]: same nit about this change

**Review by marcelofabri — APPROVED** [review]
This is great! Can you please also rebase before merging as some rules got merged after you created your branch? 🙏

File: Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift (hunk 3 above)
Comment (SimplyDanny) [reply Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift:63]: I already had it in there but cannot remember why I decided to leave it out at the end.

Could you mention an example rule for which this would be useful?

Comment (marcelofabri) [reply Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift:63]: any of the "Count" violations, like `FunctionParameterCountRule`

**Review by marcelofabri — APPROVED** [review]
